server/src/main/java/io/crate/metadata/upgrade/IndexTemplateUpgrader.java

```diff
@@ -114,4 +122,48 @@ private HashMap<String, IndexTemplateMetadata> archiveUnknownOrInvalidSettings(M
         }
         return upgradedTemplates;
     }
+
+    public static boolean populateColumnPositions(Map<String, Object> mapping) {
+        var columnPositionResolver = new ColumnPositionResolver<Map<String, Object>>();
+        populateColumnPositions("", mapping, 1, columnPositionResolver, new HashSet<>());
+        ColumnPositionResolver.resolve(columnPositionResolver);
+        return columnPositionResolver.numberOfColumnsToReposition() > 0;
+    }
+
+    private static void populateColumnPositions(String parentName,
+                                                Map<String, Object> mapping,
+                                                int currentDepth,
+                                                ColumnPositionResolver<Map<String, Object>> columnPositionResolver,
+                                                Set<Integer> takenPositions) {
+
+        Map<String, Object> properties = Maps.get(mapping, "properties");
+        if (properties == null) {
+            return;
+        }
+        Map<String, Map<String, Object>> childrenColumnProperties = new TreeMap<>(Comparator.naturalOrder());
+        for (var e : properties.entrySet()) {
+            String name = parentName + e.getKey();
+            Map<String, Object> columnProperties = (Map<String, Object>) e.getValue();
+            columnProperties = furtherColumnProperties(columnProperties);
+            Integer position = (Integer) columnProperties.get("position");
+            if (position == null || takenPositions.contains(position)) {
+                columnPositionResolver.addColumnToReposition(name,
+                                                             columnProperties,
+                                                             (cp, p) -> cp.put("position", p),
+                                                             currentDepth);
+            } else {
+                takenPositions.add(position);
+                columnPositionResolver.updateMaxColumnPosition(position);
+            }
+            childrenColumnProperties.put(name, columnProperties);
+        }
+        // Breadth-First traversal
+        for (var childColumnProperties : childrenColumnProperties.entrySet()) {
+            populateColumnPositions(childColumnProperties.getKey(),
+                                    childColumnProperties.getValue(),
+                                    currentDepth + 1,
+                                    columnPositionResolver,
+                                    takenPositions);
+        }
+    }
```

Comment:
this helps resolve duplicate positions.
when a top level column and an inner column both took the same position, breadth-first order is needed.